jkube-kit/build/service/docker/src/test/java/org/eclipse/jkube/kit/build/service/docker/RegistryServicePullIT.java

```diff
@@ -0,0 +1,244 @@
+import static org.mockito.Mockito.verify;
+
+class RegistryServicePullIT {
+  private DefaultMockServer server;
+  private KitLogger kitLogger;
+
+  @BeforeEach
+  void setUp() {
+    kitLogger = spy(new KitLogger.SilentLogger());
+    server = new DefaultMockServer();
+    server.start();
+  }
+
+  @AfterEach
+  void tearDown() {
+    server.shutdown();
+  }
+
+  @ParameterizedTest(name = "pull {0} and no registry from any source, then pull url = {1}")
+  @CsvSource({
+      "word:word,/v1.18/images/create?fromImage=word&tag=word",
+      "word/word:tag,/v1.18/images/create?fromImage=word%2Fword&tag=tag",
+      "word.word/word/word:tag,/v1.18/images/create?fromImage=word.word%2Fword%2Fword&tag=tag",
+      "word:5000/word:tag,/v1.18/images/create?fromImage=word%3A5000%2Fword&tag=tag",
+      "word.word:5000/word:tag,/v1.18/images/create?fromImage=word.word%3A5000%2Fword&tag=tag",
+      "word.word/word:tag,/v1.18/images/create?fromImage=word.word%2Fword&tag=tag",
+      "word.word/word.word/word:tag,/v1.18/images/create?fromImage=word.word%2Fword.word%2Fword&tag=tag",
+      "word.word.word/word:tag,/v1.18/images/create?fromImage=word.word.word%2Fword&tag=tag",
+      "word.word.word/word/word:tag,/v1.18/images/create?fromImage=word.word.word%2Fword%2Fword&tag=tag"
+  })
+  void pullImageWithPolicy_whenNoRegistryFromAnySource_thenPullImage(String imageName, String pullImageEndpoint) throws IOException {
+    // Given
+    server.expect().post()
+        .withPath(pullImageEndpoint)
+        .andReturn(200, Collections.singletonMap("message", "OK"))
+        .once();
+    RegistryService registryService = createNewRegistryService(server.getPort());
+    ImagePullManager imagePullManager = ImagePullManager.createImagePullManager("Always", "true", new Properties());
+    BuildConfiguration buildConfiguration = BuildConfiguration.builder().build();
+    RegistryConfig registryConfig = createNewRegistryConfig(null);
+
+    // When
+    registryService.pullImageWithPolicy(imageName, imagePullManager, registryConfig, buildConfiguration);
+
+    // Then
+    verify(kitLogger).info(eq("Pulled %s in %s"), eq(imageName), anyString());
+  }
+
+  @ParameterizedTest(name = "pull {0} and registry from registry config, then registry not prepended to image name {1}")
+  @CsvSource({
+      "word.word/word/word:tag,/v1.18/images/create?fromImage=word.word%2Fword%2Fword&tag=tag",
+      "word:5000/word:tag,/v1.18/images/create?fromImage=word%3A5000%2Fword&tag=tag",
+      "word.word:5000/word:tag,/v1.18/images/create?fromImage=word.word%3A5000%2Fword&tag=tag",
+      "word.word/word.word/word:tag,/v1.18/images/create?fromImage=word.word%2Fword.word%2Fword&tag=tag",
+      "word.word.word/word/word:tag,/v1.18/images/create?fromImage=word.word.word%2Fword%2Fword&tag=tag"
+  })
+  void pullImageWithPolicy_whenRegistryFromRegistryConfig_thenRegistryNotPrependedToImageName(String imageName, String pullImageEndpoint) throws IOException {
```

Comment:
Given the description, I don't fully understand why in this test the registry is not prepended to the image name while in the next test `pullImageWithPolicy_whenRegistryFromRegistryConfig_thenRegistryPrependedToImageName` which has almost the same description but with a different expected outcome, the registry is prepended.

Reply:
I've updated the description to:
```
@ParameterizedTest(name = "pull {0} image name already contains registry and registry from registry config, then registry not prepended to image name {1}")
void pullImageWithPolicy_whenImageNameAlreadyContainsRegistryAndRegistryFromRegistryConfig_thenRegistryNotPrependedToImageName
```

Reply:
Still not following. Isn't the registry quay.io?

Reply:
In this test when registry is already present in image name, then registry configuration would have no effect. Precedence would always be given to registry in image name (if present). For example:

- word.word/word/word:tag, `word.word` would be considered registry
- word:5000/word:tag, `word:5000` would be considered registry
- word.word:5000/word:tag, `word.word:5000` would be considered registry
- word.word/word.word/word:tag, `word.word` would be considered registry
- word.word.word/word/word:tag, `word.word.word` would be considered registry

Reply:
That's not clearly stated in the description then.
`image name already contains registry and registry from registry config`
this is confusing.
I assume that what we're saying is that the image name contains the registry information and that this one takes precedence over the one provided in the registry configuration

Reply:
yes, something like that. When doing this stuff, try to keep in mind the Rohan from the future and how you would like to help him get to understand this as quick as possible